msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/mwpj/TestCase2521946.kt

```diff
@@ -0,0 +1,141 @@
+import com.microsoft.identity.client.ui.automation.interaction.PromptParameter
+import com.microsoft.identity.labapi.utilities.client.ILabAccount
+import com.microsoft.identity.labapi.utilities.client.LabQuery
+import com.microsoft.identity.labapi.utilities.constants.AzureEnvironment
+import com.microsoft.identity.labapi.utilities.constants.TempUserType
+import com.microsoft.identity.labapi.utilities.constants.UserType
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+
+// https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2521946
+// [MWPJ] Device registration entry migration (same upn)
+@SupportedBrokers(brokers = [BrokerHost::class])
+@LocalBrokerHostDebugUiTest
+class TestCase2521946 : AbstractMsalBrokerTest() {
+    private lateinit var mUsGovAccount: ILabAccount
+    private lateinit var mBrokerHostApp: BrokerHost
+
+    @Test
+    fun test_2521946() {
+        // Register 2 accounts from different tenants
+        mBrokerHostApp.multipleWpjApiFragment.performDeviceRegistration(mUsGovAccount.username, mUsGovAccount.password)
+        mBrokerHostApp.multipleWpjApiFragment.performDeviceRegistration(mLabAccount.username, mLabAccount.password)
+        val deviceRegistrationRecords = mBrokerHostApp.multipleWpjApiFragment.allRecords
+        Assert.assertEquals(2, deviceRegistrationRecords.size)
+
+        // Unregister the device from the legacy space
+        mBrokerHostApp.multipleWpjApiFragment.unregister(mUsGovAccount.username)
+
+        // Verify that the device is unregistered for the legacy API
+        val errorMessage = mBrokerHostApp.accountUpn
+        Assert.assertNotNull(errorMessage)
+        Assert.assertTrue(errorMessage!!.contains("Device is not Workplace Joined"))
+
+        // Verify that the device is still registered for the second account using the MWPJ API.
+        val recordInExtendedSpace = mBrokerHostApp.multipleWpjApiFragment.getRecordByUpn(mLabAccount.username)
+        Assert.assertNotNull(recordInExtendedSpace)
+
+        // re-register the device with the same account using the legacy API
+        val promptHandlerParameters = PromptHandlerParameters.builder()
+                .prompt(PromptParameter.WHEN_REQUIRED)
+                .consentPageExpected(false)
+                .passwordPageExpected(false)
+                .sessionExpected(true)
+                .loginHint(mLabAccount.username)
+                .build()
+
+        mBrokerHostApp.performDeviceRegistration(mLabAccount.username, mLabAccount.password, promptHandlerParameters)
+
+        // Verify the device is registered for the legacy API
+        val legacyAccountMessage = mBrokerHostApp.accountUpn
+        Assert.assertNotNull(legacyAccountMessage)
+        Assert.assertTrue(legacyAccountMessage!!.contains(mLabAccount.username))
+
+        // Verify the entry is the same, it was just migrated
+        val recordInLegacy = mBrokerHostApp.multipleWpjApiFragment.getRecordByUpn(mLabAccount.username)
```

Comment:
we want to check that *only* one record is return. and that only record must match.

This getRecordByUpn doesn't seem to check that (you could get >1 accounts, and then the method will filter it later.)